Feature: Partial Parsing

beckjake opened the pull request:
First implementation of the "Partial parsing" feature

This PR implements "partial parsing", which writes a pickle file to the target directory and, if `--partial-parse` is passed on the command line, uses it for future runs to skip parsing nodes. It attempts to be smart and only pick nodes with files that have the same hash.

I refactored parsing a lot (ok, basically rewrote it) to make it multi-stage so it knows about files and such before it parses, and emits results into a common object.

It invalidates the parsed file in certain conditions:
 - changes to `--vars` on the command-line between runs
 - changes to `profiles.yml` between runs
 - changes to `dbt_project.yml` between runs (for _any_ projects/dependencies)
 - changes to `--profile` or `--target` parameters

Tricky things:
 - For seeds, the "hash" is their filepath, since contents have no reflection on parsing
 - If you move your project at all, that invalidates the cache, because the cache keys are absolute paths (because I live in eternal fear of bad cache data)
 - We now parse per-project rather than per-type for everything but macros, which I think is ok! It makes it a bit easier to do cache management for the intermediate parsing stuff, and I was able to be a bit more conservative on caching (we shouldn't improperly cache across projects that might share abspaths for whatever evil symlink-y reason)
 - If you change your env_vars and it would change parse results, we won't detect it and invalidate the cache, so you will probably be sad
 - I deleted a lot of unit tests, but they weren't very good tests, nor very unit. Not that mine are tons better, but they do pass and should be less brittle.

Unrelated but actually related changes here:
 - I changed some aspects of fqns that might cause problems. They will definitely change `dbt ls --output selector` output:
   - test FQNs/unique IDs are in my opinion much more reasonable. Before, calling a test in another package in your schema.yml with the `other_pkg.my_test` syntax would result in a test that had a package name of `other_pkg` instead of the current project. FQNs were similar. Instead, `other_pkg` is now part of the test name.
   - snapshots now get their filename as part of the FQN. This will be more useful when we implement models as blocks, but it seems like otherwise you lose config selectivity for putting multiple blocks in one file. I only did it the way I did it before because I did not understand parsing... no excuses now.
 - I moved `database` and `schema` into the base Credentials class. We had a shadow dependency upon all `Credentials` having those values and it came up during testing.
 - I had to move `SourceConfig` out of the `parser/` directory to handle weird import ordering things.
 - I moved some validation calls behind the `--strict` flag because they were taking up an absurd amount of time (multiple seconds)

Things that came along for the ride:
 - I verified stuff in the `parser/` directory and `loader.py` using mypy, so there are a lot of type annotations 
   - I also added a `third-party-stubs/` folder with some basic agate.Table stuff
 - I bumped dbt to 0.15.0a1 - this should be easy to revert if we want since I left it as its own commit, but switching back and forth between `dev/0.14.1` and knowing where I was during testing was pretty annoying
 - I added a `--no-write-json` flag, with it dbt no longer writes manifest.json, run_results.json, or graph.gpickle. These take a huge amount of time on larger projects!

With these changes, a `dbt --no-write-json --partial-parse ls` on a 100-node/300-test project that used to take 4s now takes 1s. The 1000/3000 project went from ~40s to ~5s. `dbt compile` saw much less improvement since we still compile everything - you'll never do better than a 50% speedup there.

Still TODO:
  - tests that actually exercise partial parsing itself!
  - handle source code TODOs I left
  - (future PR): ParserUtils is now pretty slow on large projects because ref/source/doc resolution has (I think) some unpleasant quadratic-or-worse characteristics. On the 1000/3000 project it's >80% of runtime if the cache is available and you don't write files.

cmcarthur commented on core/dbt/contracts/graph/manifest.py:
```diff
@@ -1,23 +1,174 @@
+        checksum = FileHash.from_contents(file_contents)
+        self = cls(path=path, checksum=checksum)
+        self._contents = file_contents.strip()
+        return self
```
this is the only thing i take issue with architecturally in this whole PR. we should keep business logic and I/O separate. introducing a dependency from our pure business logic world (dataclasses) onto our pure I/O world (clients.system) is a bad idea.

IMO this function should take a second arg which is either `file_contents: str` or the more pythonic and possibly more performant `file_handle: typing.TextIO`. then you should push the actual reading of the file into the parser.

in general, i'd prefer the `contracts` namespace to not rely on other namespaces that perform I/O. it makes them more pure, easier to test, and easier to reason about.

beckjake replied:
Sure. I don't really agree about that (I think pushing this into parsing makes it harder to follow what's going on and easier to mess up) but if it's important to you I'm happy to change it.

cmcarthur commented on core/dbt/contracts/graph/manifest.py:
```diff
@@ -1,23 +1,174 @@
+                os.path.join(root, searched_path, relative_path)
+            )
+        )
+        return cls.from_file(match)
```
i think this is unused

beckjake commented on core/dbt/__init__.py:
```diff
@@ -1 +0,0 @@
-__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
we do not! [PEP 420](https://www.python.org/dev/peps/pep-0420/) was introduced in Python 3.3.

beckjake commented on core/dbt/parser/results.py:
```diff
@@ -0,0 +1,191 @@
+    macros: Dict[str, ParsedMacro] = field(default_factory=dict)
+    patches: Dict[str, ParsedNodePatch] = field(default_factory=dict)
+    files: Dict[str, SourceFile] = field(default_factory=dict)
+    disabled: Dict[str, List[ParsedNode]] = field(default_factory=dict)
```
This needs to have information that `Manifest` does not, and is missing information that `Manifest` has - the possibilities are more constrained, as you can't have compiled entities, and we keep track of file paths to the various disabled nodes instead of a flattened list. I thought about making it a subclass but then there were annoying type issues

beckjake commented on core/dbt/task/debug.py:
```diff
@@ -211,6 +211,7 @@ def _load_profile(self):
         try:
             self.profile = Profile.from_args(self.args, self.profile_name)
         except dbt.exceptions.DbtConfigError as exc:
+            raise
```
No! thank you for catching this :(

beckjake commented on core/dbt/loader.py:
```diff
@@ -1,163 +1,240 @@
+        '\0'.join([
+            getattr(config.args, 'vars', '{}') or '{}',
+            getattr(config.args, 'profile', '') or '',
+            getattr(config.args, 'target', '') or '',
```
`os.environ` triggers almost everything, sadly: There are a bunch of environment variables on various oses that change when you open a new terminal, log out + in, etc. I'll add dbt version though, that's a good one.

beckjake commented:
>one final thought: the mypy work looks great. can you wire that up in the circleci build so we keep mypy happy?

Keeping mypy happy implies it's happy right now, but it's actually mad about two lines that I had been reluctant to deal with fixing because they're kind of hard:
```
core/dbt/parser/base.py:323: error: Too few arguments
core/dbt/parser/rpc.py:34: error: Argument 1 of "get_compiled_path" is incompatible with supertype "ConfiguredParser"; supertype defines the argument type as "FileBlock"
```

I'll have to write some kind of gross things to get around the first one, the second one is just going to be a little under-specified.

drewbanin commented:
I'm going to dig into this (and do some smoke testing) today

beckjake commented on core/dbt/loader.py:
```diff
@@ -1,163 +1,243 @@
+            self.parse_project(project, macro_manifest, old_results)
+
+    def write_parse_results(self):
+        path = os.path.join(self.root_project.target_path,
```
This is relative.

beckjake commented on core/dbt/contracts/graph/manifest.py:
```diff
@@ -1,23 +1,152 @@
+
+    @property
+    def search_key(self):
+        # TODO: should this be project root + original_file_path?
```
I think so, yes. Though it doesn't really matter since the project root is always absolute.

beckjake commented:
For problem 1, it was because we parsed just the internal projects and then all of the projects, and cached each. I've refactored manifest loading a bit to fix this behavior.

For problem 2:
Previously, duplicate results silently overwrote the existing ones. The message and result is dumb because on tests, compilation errors cause the offending entry to just be skipped instead of failing the run (like duplicate models would). I think the easiest thing to do is change it back to ignoring duplicate schema tests, so that's what I've done, though it does feel a bit bad. I fixed sources too, in the same place (`raise_duplicate_resource_name`).

drewbanin commented:
- It looks like `target_database` has become a required property for Snapshots. I believe it was previously optional (defaulting to the database from the active target)
- I'm still seeing this error for duplicated sources:
```
Encountered an error:
Compilation Error
  dbt found two resources with the name "database_account". Since these resources have the same name,
  dbt will be unable to find the correct resource when source("database_account") is used. To fix this,
  change the name of one of these resources:
  - source.fishtown_internal_analytics.sinter_prod.database_account (models/staging/sinter_postgres/schema.yml)
  - source.fishtown_internal_analytics.sinter_prod.database_account (models/staging/sinter_postgres/schema.yml)
```

Which is honestly good and fine - we should just update the error message from `source('database_account')` to `source('schema_name', 'database_account')`

- Duplicate schema.yml entries warning:

The error message we show for duplicated schema.yml entries is unnecessarily verbose. Right now, it looks like
```
Encountered an error:
Compilation Error
  dbt found two schema.yml entries for the same model named client_emails.
The first patch was specified in ParsedNodePatch(name='client_emails', description='', original_file_path='models/units/base/clients/schema.yml', columns={'client_listing_email_id': ColumnInfo(name='client_listing_email_id', description='')}, docrefs=[])
and the second in ParsedNodePatch(name='client_emails', description='', original_file_path='models/units/base/schema.yml', columns={'client_listing_email_id': ColumnInfo(name='client_listing_email_id', description='')}, docrefs=[]).
Models and their associated columns may only be described a single time.
```

Can we change this to:
```
  dbt found two schema.yml entries for the same model named client_emails.
Models and their associated columns may only be described a single time.
To fix this, remove the model entry for for client_emails in one of these files:
 - models/units/base/schema.yml
 - models/units/base/clients/schema.yml
```

beckjake commented:
>It looks like target_database has become a required property for Snapshots. I believe it was previously optional (defaulting to the database from the active target)

I don't think that's true, because we have a test asserting that we fail without a database in `dev/0.14.1`:
[a test case](https://github.com/fishtown-analytics/dbt/blob/b2f2e69377dc9e9082b61358cc508b4a947e573b/test/integration/004_simple_snapshot_test/test_simple_snapshot.py#L336) and its [contents](
https://github.com/fishtown-analytics/dbt/blob/b2f2e69377dc9e9082b61358cc508b4a947e573b/test/integration/004_simple_snapshot_test/test-snapshots-invalid/snapshot.sql)

drewbanin commented:
I think I changed that test to make `target_database` optional :) 

This test fails because `target_schema` is not specified -- I did a bad job updating the snapshot name there. See this commit: https://github.com/fishtown-analytics/dbt/commit/08d79cc324918e9ada44ec7864871c850995b9d2#diff-62e19c87afca47b16a0128d8da31a914R10
